services/src/main/java/org/keycloak/services/clientregistration/policy/impl/ProtocolMappersClientRegistrationPolicy.java

```diff
@@ -48,27 +51,51 @@ public ProtocolMappersClientRegistrationPolicy(KeycloakSession session, Componen
 
     @Override
     public void beforeRegister(ClientRegistrationContext context) throws ClientRegistrationPolicyException {
-        testMappers(context);
+        testMappers(context, null);
     }
 
-    protected void testMappers(ClientRegistrationContext context) throws ClientRegistrationPolicyException {
+    protected void testMappers(ClientRegistrationContext context, ClientModel clientModel) throws ClientRegistrationPolicyException {
         List<ProtocolMapperRepresentation> protocolMappers = context.getClient().getProtocolMappers();
         if (protocolMappers == null) {
             return;
         }
 
         List<String> allowedMapperProviders = getAllowedMapperProviders();
 
-        for (ProtocolMapperRepresentation mapper : protocolMappers) {
-            String mapperType = mapper.getProtocolMapper();
-
-            if (!allowedMapperProviders.contains(mapperType)) {
-                ServicesLogger.LOGGER.clientRegistrationMapperNotAllowed(mapper.getName(), mapperType);
-                throw new ClientRegistrationPolicyException("ProtocolMapper type not allowed");
-            }
-        }
+        for (ProtocolMapperRepresentation mapperRepresentation : protocolMappers) {
+            String mapperType = mapperRepresentation.getProtocolMapper();
+
+			if (allowedMapperProviders.contains(mapperType)) {
+				continue;
+			}
+			if (clientModel == null) {
+				failWithProtocolMapperTypeNotAllowedError(mapperRepresentation);
+				return;
+			}
+			String mapperRepresentationId = mapperRepresentation.getId();
+			if (mapperRepresentationId == null) {
+				failWithProtocolMapperTypeNotAllowedError(mapperRepresentation);
```

Comment:
We can probably respond with a more explicit message that the mapper id is missing and later that it can not be found?

Instead of just returning `ProtocolMapper type not allowed`?